server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala

```diff
@@ -0,0 +1,182 @@
+      "Upgrade".equalsIgnoreCase(req.headers().get(HttpHeaderNames.CONNECTION))
+
+  override def handlerAdded(ctx: ChannelHandlerContext): Unit =
+    if (ctx.channel.isActive) {
+      initHandler(ctx)
+    }
+  override def channelActive(ctx: ChannelHandlerContext): Unit = {
+    channelGroup.add(ctx.channel)
+    initHandler(ctx)
+  }
+
+  private[this] def initHandler(ctx: ChannelHandlerContext): Unit = {
+    if (eventLoopContext == null)
+      eventLoopContext = ExecutionContext.fromExecutor(ctx.channel.eventLoop)
+  }
+
+  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
+    logger.error("Error while processing the request", cause)
+  }
+
+  override def channelRead(ctx: ChannelHandlerContext, msg: Any): Unit = {
+    def replyWithError500(reason: Throwable): Unit = {
+      logger.error("Error while processing the request", reason)
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.INTERNAL_SERVER_ERROR)
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, 0)
+      res.headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE)
+      val _ = ctx.writeAndFlush(res)
+    }
+
+    def rejectHandshakeForRegularEndpoint(content: NettyResponseContent): Unit = {
+      val message = "Unexpected WebSocket handhake on a regular HTTP endpoint"
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.BAD_REQUEST, Unpooled.wrappedBuffer(message.getBytes))
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, message.length())
+      res.headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE)
+      content.channelPromise.setFailure(new IllegalStateException("Unexpected response content"))
+      val _ = ctx.writeAndFlush(res)
+    }
+
+    msg match {
+      case req: FullHttpRequest if isWsHandshake(req) =>
+        ctx.pipeline().remove(this)
+        ctx.pipeline().remove(classOf[ReadTimeoutHandler])
+        ReferenceCountUtil.release(msg)
+        val (runningFuture, _) = unsafeRunAsync { () =>
+          route(NettyServerRequest(req.retain()))
+            .map {
+              case Some(response) => response
+              case None           => ServerResponse.notFound
+            }
+        }
+
+        val _ = runningFuture.transform {
+          case Success(serverResponse) =>
+            try {
+              serverResponse.body match {
+                case Some(function) => {
+                  val content = function(ctx)
+                  content match {
+                    case r: ReactiveWebSocketProcessorNettyResponseContent =>
+                      initWsPipeline(ctx, r, req)
+                    case otherContent =>
+                      rejectHandshakeForRegularEndpoint(otherContent)
+                  }
+                }
+                case None =>
+                  replyWithError500(new IllegalArgumentException("Missing response body, expected WebSocketProcessorNettyResponseContent"))
```

Comment:
doesn't the lack of a body simply  mean that the endpoint was a "normal" one, but didn't specify a body output? So, is this a server error?

Reply:
Yes, I just checked and an empty body endpoint would trigger this. What's worse, a 404 would also end up here. I'll fix this.